lib/src/settings.rs

```diff
@@ -241,3 +245,137 @@ impl<T> ConfigResultExt<T> for Result<T, config::ConfigError> {
         }
     }
 }
+
+/// A size in bytes optionally formatted/serialized with binary prefixes
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct HumanByteSize(pub u64);
+
+impl std::fmt::Display for HumanByteSize {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (value, prefix, _) = binary_prefix(self.0 as f32);
+        write!(f, "{value}{prefix}B")
+    }
+}
+
+impl serde::Serialize for HumanByteSize {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        if !serializer.is_human_readable() {
+            return self.0.serialize(serializer);
+        }
+        // Use a readable format if the value is exactly representable with an integer
+        // and a unit; otherwise, use bytes
+        let (value, prefix, exponent) = binary_prefix(self.0 as f32);
+        let value = value as u64;
+        if exponent > 0 && value * 1024u64.pow(exponent) == self.0 {
+            let s = format!("{value}{prefix}B");
+            s.serialize(serializer)
+        } else {
+            self.0.serialize(serializer)
+        }
```

Comment:
It can also be modeled as untagged enum + `HumanByteSize` struct.
```rust
#[serde(untagged)]
enum MaybeHumanByteSize {
    Exact(i64),
    Human(HumanByteSize),
}

#[serde(try_from = "String")]
struct HumanByteSize { .. }
```
https://github.com/serde-rs/serde/issues/939#issuecomment-939514114

I think that's easier than implementing `Deserialize`/`Serialize`, and we don't have to think about whether the value can be represented as human-readable integer.

Reply:
On review, we don't actually need the `Serialize` impl at all. The `Deserialize` impl is boilerplatey but fundamentally simple, so my inclination is to stick with it rather than relying on more esoteric macro magic.

Reply:
`serde::Deserialize` is also magic to me, but I don't have strong opinion. With the removal of `Serialize`, we won't need a type that can preserve the serialized format. That's good.